File: day13/Day13.scala
import scala.io.Source.stdin

@main def main(idx: Int) =
  val (nums, folds) = stdin.getLines().span{_ != ""}
  val ns = nums.flatMap{_.split(",").map(_.toInt).sliding(2)
                         .map{case Array(a, b) => (a, b) }}.toSet
  val fs = folds.drop(1).flatMap{_.split("=").sliding(2)
                                  .map{case Array(a, b) => (a.last, b.toInt)}}.toList
  val List(w, h) = List('x', 'y').map{c => fs.filter{c == _._1}.map{_._2}.max * 2 + 1}

  val top = if (idx == 1) 1 else fs.length
  val (fw, fh, fns) = fs.take(top).foldLeft((w, h, ns)){(t, f) => (t, f) match {
    case ((w, h, nums), (dir, fold)) =>
      ( if (dir == 'x') w / 2 else w
      , if (dir == 'y') h / 2 else h
      , nums.map{case (x, y) => dir match {
          case 'x' if x >= fold => (w - x - 1, y)
          case 'y' if y >= fold => (x, h - y - 1)
          case _ => (x, y)
        }}
      )
  }}

  println{
    if idx == 1
      then fns.size
      else (0 until fh).map{y => (0 until fw).map{x =>
              if (fns.contains{(x, y)}) '#' else ' '}.mkString}
           .mkString("\n")
  }

File: day13/Day13.py
from sys import argv, stdin

nums = set()
folds = []

for l in map(str.strip, stdin.readlines()):
    if l.startswith("fold"):
        a, b = l[len("fold along") + 1 :].split("=")
        folds.append([a, int(b)])
    elif len(l.strip()) == 0:
        pass
    else:
        a, b = l.split(",")
        nums.add((int(a), int(b)))

w = max(n for dir, n in folds if dir == "x") * 2 + 1
h = max(n for dir, n in folds if dir == "y") * 2 + 1

if argv[1] == "1":
    top = 1
else:
    top = len(folds)

for dir, n in folds[:top]:
    new_nums = set()
    for x, y in nums:
        if dir == "y":
            if y >= n:
                new_nums.add((x, h - y - 1))
            else:
                new_nums.add((x, y))
        else:
            if x >= n:
                new_nums.add((w - x - 1, y))
            else:
                new_nums.add((x, y))
    if dir == "y":
        h = h // 2
    else:
        w = w // 2
    nums = new_nums

if argv[1] == "1":
    print(len(nums))
else:
    mat = [["#" if (x, y) in nums else " " for x in range(w)] for y in range(h)]
    print("\n".join("".join(r) for r in mat))
